### PR TITLE
Overall review

### PR DESCRIPTION
**Pull request by GregoryMaks**

### CODE REVIEW

File: RingTask/Classes/Flow/RedditFlow/Data/RedditTopListingDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    // MARK: - Private methods
+    
+    private func loadDataChunk(loadingType: LoadingType) {
+        guard !isLoading else { return }
+        
+        let pagingMarker: RedditPagingMarker? = (loadingType == .loadMore) ? self.pagingMarker : nil
+        
+        isLoading = true
+        redditService.requestTopPosts(pagingMarker: pagingMarker) { [weak self] result in
+            result.analysis(ifValue:
```
Comment (atmamont) [inline RingTask/Classes/Flow/RedditFlow/Data/RedditTopListingDataSource.swift]: Can't decide if it really better than 
```
switch result {
case .sucess(let response):
case .failure(let error):
}
```
it seems event to be more readable without closures

File: RingTask/Classes/Flow/RedditFlow/Data/RedditTopListingDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                { response in
+                    guard let strongSelf = self else { return }
+                    
+                    strongSelf.modifyModels(withNewData: response.models, append: (loadingType == .loadMore))
```
Comment (atmamont) [inline RingTask/Classes/Flow/RedditFlow/Data/RedditTopListingDataSource.swift]: `updateModels`, `shouldAppend`

File: RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+//
+//  Created by Hryhorii Maksiuk on 8/7/17.
+//  Copyright © 2017 Gregory M. All rights reserved.
+//
+
+import UIKit
+import SafariServices
+
+
+class RedditFlowCoordinator: NSObject {
```
Comment (atmamont) [inline RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift]: WTF at first look. Then it become clear it is required by SafariVC. Can it be conformed to `NSObject` in appropriate extension?

File: RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        self.window = window
+        self.networkService = networkService
+        self.imageLoadingService = imageLoadingService
+    }
+    
+    func start() {
+        let storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
+        guard let navController = storyboard.instantiateInitialViewController() as? UINavigationController else {
+            print("Error, root navigation not found")
+            return
```
Comment (atmamont) [inline RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift]: what user will see in this case?
IMO better to `fatalError` here

File: RingTask/Classes/Flow/RedditFlow/RedditTopListingViewController.swift [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+                                      message: nil,
+                                      preferredStyle: .actionSheet)
+        alert.addAction(UIAlertAction(title: "Save attached image to gallery", style: .default) { [weak self] _ in
+            guard let strongSelf = self else { return }
+            
+            strongSelf.showFullScreenLoadingIndicator()
+            strongSelf.viewModel.saveImageToGallery(for: itemModel) { result in
+                strongSelf.hideFullScreenLoadingIndicator(animated: false) {
+                    
+                    result.map(ifSuccess: { strongSelf.showFullscreenMessage("Your image was saved", duration: 2.0) },
```
Comment (atmamont) [inline RingTask/Classes/Flow/RedditFlow/RedditTopListingViewController.swift]: HOLY MOLY! :)

File: RingTask/Classes/Helpers/AuthorDateFormatter.swift [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import Foundation
+
+
+class AuthorDateFormatter {
+    
+    lazy var calendar: Calendar = {
+        return Calendar(identifier: Calendar.Identifier.gregorian)
+    }()
+    
+    func authorDateString(forAuthor author: String, date: Date) -> String {
```
Comment (atmamont) [inline RingTask/Classes/Helpers/AuthorDateFormatter.swift]: `func stringValue(forAuthor: String...` ?

File: RingTask/Classes/Networking/NetworkService.swift [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+enum NetworkError: Error {
+    
+    case unknown
+    case timeOut
+    case networkConnectionLost
+    case cancelled
+    case other(urlError: Error)
+    
+    init(urlError: Error?) {
+        // TODO
```
Comment (atmamont) [inline RingTask/Classes/Networking/NetworkService.swift]: better to remove //TODO

File: RingTask/Classes/Utilities/Currying.swift [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+public func curry<A, B>(_ function: @escaping (A) -> B) -> (A) -> B {
+    return { (a: A) -> B in function(a) }
+}
+
+public func curry<A, B, C>(_ function: @escaping (A, B) -> C) -> (A) -> (B) -> C {
+    return { (a: A) -> (B) -> C in { (b: B) -> C in function(a, b) } }
+}
+
+public func curry<A, B, C, D>(_ function: @escaping (A, B, C) -> D) -> (A) -> (B) -> (C) -> D {
```
Comment (atmamont) [inline RingTask/Classes/Utilities/Currying.swift]: 🥇

File: RingTask/Classes/Utilities/Result.swift [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        case let .failure(error): return ifError(error)
+        }
+    }
+    
+    // MARK: Raw value related
+    
+    @discardableResult func map<T, E>(ifSuccess: @escaping (TValue) -> T, ifFailure: @escaping (TError) -> E)
+        -> Result<T, E>
+    {
+        return self.flatMap(ifSuccess: ifSuccess • liftValue, ifFailure: ifFailure • liftError)
```
Comment (atmamont) [inline RingTask/Classes/Utilities/Result.swift:76]: alt text

**Review by atmamont — COMMENTED** [review]
Code readability is great on top level but is fucking weird and totally unreadable inside. Some kind of inner empire :)
Overall feeling - 👍

File: RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift (hunk 3 above)
Comment (GregoryMaks) [reply RingTask/Classes/Flow/RedditFlow/RedditFlowCoordinator.swift]: Unfortunately no, you cannot inherit from class in extension, but I've added a comment! )

File: RingTask/Classes/Flow/RedditFlow/RedditTopListingViewController.swift (hunk 5 above)
Comment (GregoryMaks) [reply RingTask/Classes/Flow/RedditFlow/RedditTopListingViewController.swift]: that's good or bad? :D

File: RingTask/Classes/Networking/NetworkService.swift (hunk 7 above)
Comment (GregoryMaks) [reply RingTask/Classes/Networking/NetworkService.swift]: removed and improved

File: RingTask/Classes/Utilities/Currying.swift (hunk 8 above)
Comment (GregoryMaks) [reply RingTask/Classes/Utilities/Currying.swift]: unfortunately didn't use anywhere, so had to remove :(

File: RingTask/Classes/Utilities/Result.swift (hunk 9 above)
Comment (GregoryMaks) [reply RingTask/Classes/Utilities/Result.swift:76]: 🚀

**Pull request closed without merging**